### PR TITLE
Re-add the list ping compatibility checker

### PR DESCRIPTION
**Pull request by JoJoDeveloping**
This adds back the icon showing whether the server is likely to support the client before the client even joins, in the server list / multiplayer gui.

It is assumed a server is compatible if
- The registries contain the same keys
- Mods using network channels accept the other's version
- The [FMLNETVERSION](https://github.com/MinecraftForge/MinecraftForge/blob/1.13.x/src/main/java/net/minecraftforge/fml/network/FMLNetworkConstants.java#L34) is the same

It also adds a locking mechanism to the NetworkRegistry to prevent mods from adding channels on-the-fly as that might lead to bad behavior. If it's not wanted I can undo this.

Also, we send a hash of the set of all the keys in the registries instead of the keys themselves.

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/network/NetworkRegistry.java [hunk 1]
```diff
@@ -116,6 +122,10 @@ public static EventNetworkChannel newEventChannel(final ResourceLocation name, S
      */
     private static NetworkInstance createInstance(ResourceLocation name, Supplier<String> networkProtocolVersion, Predicate<String> clientAcceptedVersions, Predicate<String> serverAcceptedVersions)
     {
+        if(lock && !name.getNamespace().equals("fml")){
```
Comment (JoJoDeveloping) [inline src/main/java/net/minecraftforge/fml/network/NetworkRegistry.java:128]: FML is naughty and registers it's channels when connecting to a server the first time / accepting a client the first time.
So it's exempt from the lock.

File: src/main/java/net/minecraftforge/fml/client/ClientHooks.java [hunk 2]
```diff
@@ -76,108 +68,78 @@
...
-            JsonArray modDataArray = jsonData.get("modList").getAsJsonArray();
-            boolean moddedClientAllowed = !jsonData.has("clientModsAllowed") || jsonData.get("clientModsAllowed").getAsBoolean();
-            ImmutableMap.Builder<String, String> modListBldr = ImmutableMap.builder();
-            for (JsonElement obj : modDataArray)
-            {
-                JsonObject modObj = obj.getAsJsonObject();
-                modListBldr.put(modObj.get("modid").getAsString(), modObj.get("version").getAsString());
-            }
+            boolean b = NetworkRegistry.checkListPingCompatibilityForClient(packet.getForgeData().getPresentMods())
+                    && difference.areEqual()
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/client/ClientHooks.java]: I believe it is intended to allow connections from clients with registries that are a superset of the registries on the server, so this check should be modified to support that.

**Review by mezz — CHANGES_REQUESTED** [review]
I'm also a bit concerned that this is going to be a super large ping. It's one thing to send all the registry data when you connect, but it could be a bit much to receive registry hashes from a bunch of servers at once. Can you run some numbers to see what kind of data size we'd be expecting from these pings with a common kitchen-sink-style pack?

File: src/main/java/net/minecraftforge/fml/client/ClientHooks.java (hunk 2 above)
Comment (JoJoDeveloping) [reply src/main/java/net/minecraftforge/fml/client/ClientHooks.java]: I'm aware that sending the complete registry during list ping is likely a lot of data, which is why I [intend to] send the hash of the set of all registry keys (Set#hashCode), which is an int *per registry*. The actual data send looks like this (minus formatting):
```{"description":{"text":"A Minecraft Server"},
"players":{"max":20,"online":0},
"version":{"name":"1.13.2","protocol":404},
"forgeData":{
    "mods":[],
    "registryKeys":[
        {"namespace":"forge","path":"moddimensions","hash":0},
        {"namespace":"minecraft","path":"potiontypes","hash":196815442},
        {"namespace":"minecraft","path":"entities","hash":296289830},
        {"namespace":"minecraft","path":"biomes","hash":-587414690},
        {"namespace":"minecraft","path":"potions","hash":2094714475},
        {"namespace":"minecraft","path":"villagerprofessions","hash":-1709168725},
        {"namespace":"minecraft","path":"blocks","hash":-1813164851},
        {"namespace":"minecraft","path":"items","hash":413776779},
        {"namespace":"minecraft","path":"enchantments","hash":749676857},
        {"namespace":"minecraft","path":"tileentities","hash":-779590212},
        {"namespace":"minecraft","path":"soundevents","hash":-1298340753}],
    "numberOfMods":1,
    "fmlNetworkVersion":2}}
```
What is an actual problem is that the client registry may indeed be a superset of the server registry, and this method does not allow for checking that.. I asked before implementing and was told the contents have to match, but apparently misinterpreted 'match' as 'be equal'..
So should I remove the gameregistry comparsion?

**Comment by JoJoDeveloping** [comment]
So since you were confused in your livestream @cpw here's some explaination on the channels fields:
Channels have a resource location name, which I, to avoid splitting on ":" later, save as two json fields - namespace and path. So a channel like "fml:handshake" would have {"namespace":"fml", "path":"handshake"}.
Then version is the version. And canBeAbsent is there so that if a channel is there on the server, but the server does not require it on the client, the canBeAbsent field is true so that the client can figure and does not display the server as incompatible if that (not actually required) channel is missing on its side.

So a channel built like this
 ```
        ChannelBuilder.named(new ResourceLocation("lol", "cat")).
                networkProtocolVersion(() -> "0.1.2").
                clientAcceptedVersions(s -> s.equals("0.1.2")).
                serverAcceptedVersions(s -> s.equals("0.1.2")).
                simpleChannel()
```
leads to `"mods":[{"namespace":"lol","path":"cat","version":"0.1.2","mayBeAbsent":false}]`, whereas
```
                clientAcceptedVersions(s -> true).
```
leads to `"mods":[{"namespace":"lol","path":"cat","version":"0.1.2","mayBeAbsent":true}]`

Also make sure to have all the json fields names the same on serialise and deserialise.

**Pull request merged**